README.rst
===============================
Dev branch forked from the uArm-Developer/pyuarm tree
===============================
===============================
pyuarm --- Drivers for the 4-DOF robot arms from ufactory.cc
===============================

## Branched off pyuarm-2.4.x

Overview
========

Latest features:
- Added firmware detection to get_uarm() to allow this code to be used to talk to both uArm Metal and uArm Swift devices. See protocol.py and swift_protocol.py and uarm.py for most of the details about this. 
- method variables added to allow for continued expansion of the package to other future USB VIDs and uArm firmware. 

- I've otherwise tried to mess with the code as little as possible since the ROS uarm package uses this as its driver interface to the uArm device. 
- Tested successfully with a uArm Swift and uArm Metal running Ubuntu 14.04 using ROS Indigo. 

Feedback/testing/PRs, etc. requested, Advice and assistance gladly given, but safety requires avoiding unnecessary conversation. 

Other pages (online)

- `project page on Github`_
- `Download Page`_ with releases
- This page, when viewed online is at https://pyuarm.readthedocs.io.

Features
========
- Auto uArm Port detection
- Provide firmware_helper, firmware upgrade online
- Provide Auto Calibration tool

Requirements
============
- Python 2.7x or Python 3.4x above
- uArmProtocol Firmware (Please use ``uarmcli firmware -d`` or ``python -m pyuarm.tools.firmware -d`` to upgrade your firmware)

Installation
============

pyuarm
------
This install a package that can be used from Python (``import pyuarm``).

To install for all users on the system, administrator rights (root) may be required.

From PyPI
~~~~~~~~~
pyuarm can be installed from PyPI, either manually downloading the files and installing as described below or using::

    pip install pyuarm

From source (tar.gz or checkout)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Download the archive from http://pypi.python.org/pypi/pyuarm.
Unpack the archive, enter the ``pyuarm-x.y`` directory and run::

    python setup.py install

.. _`project page on GitHub`: https://github.com/uArm-Developer/pyuarm
.. _`Download Page`: http://pypi.python.org/pypi/pyuarm


CHANGELOG.md
##[2.4.0.12-JD] - 2017-07-09
### Update -Steve " 'dillo" Okay

1. Split out protocol.py into protocol.py and swift_protocol.py
2. Added detect_arm.py under tools/
3. Added code to detect multiple firmware versions and status codes for uArm Metal and Swift in uarm.py
4. Workaround for bug where pyuarm.connect() deadlocks because the boot-up tokens ("@1", "@5", etc.) are flushed from the serial port when get_uarm() connects to the arm. 

##[2.4.0.12] - 2017-03-11

### Update -Alex Tan

- Firmware 2.2.1

#### Changes

1. Improve Firmware flash
2. Improve Log description

## [2.4.0.9] - 2017-03-10
### Updated -Alex Tan

- Firmware 2.2.1

# Changes
1. Rewrite threading part improve performance
2. Add Wait parameters for Set Functions.


## [2.3.0.7] - 2016-12-04
### Updated -Alex Tan

### Compatible
- Firmware 2.2

### Changes
1. Fix Bugs for Non-block Mode
2. Add Teach function

## [2.3.0.3] - 2016-12-04
### Updated -Alex Tan

### Compatible
- Firmware 2.2

### Changes
1.  Support Unblock Mode
2.  rewrite FlashFirmware & calibrate. Flash Firmware only support external avrdude, calibrate only support query
3.  Fix connect Flag support
4.  Add set_polar support


## [2.2.5.1] - 2016-11-25
### Updated -Alex Tan

### Compatible
- Firmware 2.2

### Changes
1. Support wait for set_position

## [2.2.5.1] - 2016-11-25
### Updated -Alex Tan

### Compatible
- Firmware 2.2

### Changes
1. Fix Tip Sensor return False always
2. Rewrite logger logic
3. Add Relative coordinate support
4. Use printf in Calibrate


## [2.2.4] - 2016-11-16
### Updated -Alex Tan

### Compatible
- Firmware 2.2

### Changes
1. Fix install bug


## [2.2.3] - 2016-11-16
### Updated -Alex Tan

### Compatible
- Firmware 2.2

### Changes
1. Add function check_port_plug_in (beta)
2. Get Hardware version once connect
3. Fix is_connected when uarm plug out but still return True
4. Add Set Debug for logger

## [2.2.2] - 2016-11-16
### Updated -Alex Tan

### Compatible
- Firmware 2.2

### Changes
1. Fix pyuarm.__version__
2. Fix pypi MANIFEST.in not found


## [2.2.0] - 2016-11-16
### Updated -Alex Tan

### Compatible
- Firmware 2.2

### Changes
1. Compatible with firmware 2.2 (Gcode)
2. Fix Camera Disconnect No response
3. Add Protocol Version

## [2.1.1] - 2016-10-28
### Updated -Alex Tan

### Compatible
- Firmware 2.1

### Changes
1. Compatible with firmware 2.1
2. Compatible with Python2.x and Python 3.x
3. Rewrite Calibration logic
4. Rewrite Firmware logic (Reduce 3rd-party library dependencies) Include `avrdude` on Mac/Windows/Linux
5. Rename Functions in uarm. get/set for all communication protocol.

## [2.0.7] - 2016-09-06
### Updated - Alex Tan

### Compatible
- Firmware 2.0

### Fix
1. Fix list port issue
2. Fix receive F start error

## [2.0.6] - 2016-09-06
### Updated - Alex Tan

### Compatible
- Firmware 2.0

### Fix
1. Protocol:  
- Support Raw Servo Angle, SET_ANGLE = "sAngN{}V{}", SET_RAW_ANGLE = "sSerN{}V{}", GET_RAW_ANGLE = "gSer", GET_ANGLE = "gAng"
- Buzzer add stop Duration, SET_BUZZER = "sBuzF{}T{}S{}"
- version start with "v"
2. pyuarm:  
- set Debug before any action.
- if response message start with "f", means error
- set & get EEPROM return value depend on data type
3. Calibrate:  
- Add version 1.1.0
- Compatible with 2.0 firmware
4. list_uarm:  
- get_uarm()
- get_uarm_port_cli()
5. Miniterm:
- change read_servo_angle to get_servo_angle, write_servo_angle to set_servo_angle
- Add version 0.1.3
6. Others:  
- `python -m pyuarm.version` will display the library version and support version


## [2.0.5] - 2016-08-31
### Updated - Alex Tan

### Compatible
- Firmware 2.0

### Fix
**uarm-cli**
- get_analog, get_eeprom, set_eeprom, get_serial_number

## [2.0.3] - 2016-08-10
### Updated - Alex Tan

### Compatible
- Firmware 0.9.9

### Fix
**uarm-cli**
- help topic hide welcome message
- compatible with 0.9.9
- uarm-cli add debug option
- uarm-firmware fix -d argument issue

## [2.0.2] - 2016-08-10
### Updated - Alex Tan

### Compatible
- Firmware 0.9.8b

### Fix
**uarm-cli**  
- Use colorama Compatible with Windows platform
- fix firmware error if no uarm connect
- Catch KeyboardInterrupt
- add shortcut tip when cli start, and help center


## [2.0.0] - 2016-07-13
### Updated - Alex Tan

### Compatible
- Firmware 2.0

### Changes
- Brand new Protocol

### detail -2016-07-13
- Fix timeout issue when new Serial port
- remove time.sleep(3) after new Serial port, use is_ready() function instead
- Add UArmConnectException
- add disconnect() function
- flushInput before read anything
- use readline() replace old read logic
- add validate_coordinate()
- update move_to,pump_on, pump_off, set_buzzer, is_moving, get_servo_angle, get_tip_sensor
- add write_servos_angle
- update firmware_helper.py with new Protocol
- add cmd for for control

### detail - 2016-08-09
- connect() function after open port
- cli add serial mode

### detail - 2016-08-03
- Compatible with v0.9.7b

## [1.3.27] - 2016-07-08
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- Fix calibrate issue when use GUI.

## [1.3.26] - 2016-07-05
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- Firmware helper support port argument.
- improve firmware helper in mac os x if not install avrdude.
- add console-script: uarm-firmware, uarm-calibrate, uarm-list.

## [1.3.25] - 2016-07-01
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- Move Back to use version.py due to pyinstaller encounter error

## [1.3.24] - 2016-06-30
### Updated - Alex Tan

### Compatible
- Firmware 1.7

### Fix
- Use version.json replace version.py
- Add Support firmware version in version.json, if not Compatible will raise UnsupportFirmwareVersionException
- Add argument for calibrate.py
- Disable Stretch Calibration
- Change Linear Calibration, Increasing moving points
- Merge [frankyjuang](https://github.com/frankyjuang), fix typo

## [1.3.21] - 2016-06-28
### Updated - Alex Tan

### Compatible
- Firmware 1.7.1

### Fix

pyuarm/util.py
import logging
from .version import __version__
import os
from os.path import expanduser
home_dir = os.path.join(expanduser("~"), ".uarm", "")
ua_dir = os.path.join(home_dir, "assistant")

ERROR = 2
INFO  = 1
DEBUG = 0

global pylogger

def get_default_logger(debug=False):
    if debug:
        logging_level = logging.DEBUG
    else:
        logging_level = logging.INFO
    logger = logging.getLogger('pyuarm')
    logger.setLevel(logging_level)
    formatter = logging.Formatter('%(name)s - %(levelname)s - %(message)s')
    ch = logging.StreamHandler()
    ch.setFormatter(formatter)
    ch.setLevel(logging_level)
    logger.addHandler(ch)
    return logger

def init_logger(logger):
    """
    initialize global logger
    :param debug: if True, Turn on the logger debug
    :return:
    """
    global pylogger
    pylogger = logger
    printf('pyuarm version: ' + __version__)

def close_logger():
    handlers = pylogger.handlers
    for p in handlers:
        p.close()
        pylogger.removeHandler(p)

def printf(msg, type=INFO):
    """
    global print log function
    :param msg:
    :param type:
    :return:
    """
    if type == INFO:
        pylogger.info(msg)
    elif type == DEBUG:
        pylogger.debug(msg)
    elif type == ERROR:
        pylogger.error(msg)


class UArmConnectException(Exception):
    def __init__(self, errno, message=None):
        """
        uArm Connect Exception
        :param errno: 0 Unable to connect uArm, 1 unknown firmware version, 2 unsupported uArm Frimware version
        :param message:
        """
        if message is None:
            self.message = ""
        else:
            self.message = message
        self.errno = errno
        if self.errno == 0:
            self.error = "Unable to connect uArm"
        elif self.errno == 1:
            self.error = "Unknown Firmware Version"
        elif self.errno == 2:
            self.error = "Unsupported uArm Firmware Version"
        else:
            self.error = "Not Defined Error"

    def __str__(self):
        return repr(self.error + "-" + self.message)


pyuarm/protocol.py
# parameters

## SERVO NUMBER INDEX
SERVO_BOTTOM = 0
SERVO_LEFT = 1
SERVO_RIGHT = 2
SERVO_HAND = 3

## CALIBRATION FLAG
CONFIRM_FLAG = 0x80
CALIBRATION_FLAG = 10
CALIBRATION_LINEAR_FLAG = 11
CALIBRATION_SERVO_FLAG = 12
CALIBRATION_STRETCH_FLAG = 13

## OFFSET EEPROM ADDRESS
LINEAR_INTERCEPT_START_ADDRESS = 70
LINEAR_SLOPE_START_ADDRESS = 50
OFFSET_START_ADDRESS = 30
OFFSET_STRETCH_START_ADDRESS = 20

SERIAL_NUMBER_ADDRESS = 100

## EEPROM DATA TYPE INDEX
EEPROM_DATA_TYPE_BYTE = 1
EEPROM_DATA_TYPE_INTEGER = 2
EEPROM_DATA_TYPE_FLOAT = 4

## PROTOCOL MESSAGE
# READY & Response
READY                   = "@1"
OK                      = "OK"
# Set Command
SET_POSITION            = "G0 X{} Y{} Z{} F{}"
SET_POSITION_RELATIVE   = "G204 X{} Y{} Z{} F{}"
SET_SERVO_ANGLE         = "G202 N{} V{}"
STOP_MOVING             = "G203"
SET_PUMP                = "M231 V{}"
GET_PUMP                = "P231"
SET_GRIPPER             = "M232 V{}"
SET_BUZZER              = "M210 F{} T{}"
SET_POLAR               = "G201 S{} R{} H{} F{}"
ATTACH_SERVO            = "M201 N{}"
DETACH_SERVO            = "M202 N{}"
# Get Command
GET_SIMULATION          = "M222 X{} Y{} Z{} P0"
GET_FIRMWARE_VERSION    = "P203"
GET_HARDWARE_VERSION    = "P202"
GET_COOR                = "P220"
GET_SERVO_ANGLE         = "P200"
GET_IS_MOVE             = "M200"
GET_TIP_SENSOR          = "P233"
GET_POLAR               = "P221"
GET_GRIPPER             = "P232"
GET_EEPROM              = "M211 N0 A{} T{}"
SET_EEPROM              = "M212 N0 A{} T{} V{}"
GET_ANALOG              = "P241 N{}"
GET_DIGITAL             = "P240 N{}"

# Report Command
SET_REPORT_POSITION     = "M120 V{}"
REPORT_POSITION_PREFIX  = "@3"

#Set Protcol
SET_PROTOCOL = "BAR"


pyuarm/tools/list_uarms.py
from serial.tools import list_ports
import pdb
#UARM_HWID_KEYWORD = "USB VID:PID=0403:6001"
UARM_HWIDS = ['0403:6001','2341:0042']

UARM_MODEL=""

def uarm_ports():
#    pdb.set_trace()
    uarm_ports ={}

    for i in list_ports.comports():
    #    if i.hwid[0:len(UARM_HWID_KEYWORD)] == UARM_HWID_KEYWORD:
        if i.hwid[12:21] in UARM_HWIDS:
            uarm_ports[i[0]] = i.hwid[12:21]
            print i.hwid[12:21] 
            print i.hwid
            
    return uarm_ports


def check_port_plug_in(serial_id):
    ports = list_ports.comports()
    for p in ports:
       if p.serial_number == serial_id:
           return True
    return False


def get_uarm_port_cli():
    uarm_list = uarm_ports()
    ports = uarm_ports()
    if len(ports) > 1:
        i = 1
        for port in ports:
            print ("[{}] - {}".format(i, port))
            i += 1
        port_index = input("Please Choose the uArm Port: ")
        uarm_port = ports[int(port_index) - 1]
        return uarm_port
    elif len(ports) == 1:
        return uarm_list[0]
    elif len(ports) == 0:
        return None


def get_port_property(port_name):
    for p in list_ports.comports():
        if p.device == port_name:
            return p
    return None


def main():
    """
    ::

        $ python -m pyuarm.tools.list_uarms
        /dev/cu.usbserial-A600CVS9
        1 ports found

    """
    ports = uarm_ports()
    for p in ports:
        print (p)
    print ("{0} ports found".format(len(ports)))

if __name__ == '__main__':
    main()


pyuarm/swift_protocol.py
# parameters

## SERVO NUMBER INDEX
SERVO_BOTTOM = 0
SERVO_LEFT = 1
SERVO_RIGHT = 2
SERVO_HAND = 3

## CALIBRATION FLAG
CONFIRM_FLAG = 0x80
CALIBRATION_FLAG = 10
CALIBRATION_LINEAR_FLAG = 11
CALIBRATION_SERVO_FLAG = 12
CALIBRATION_STRETCH_FLAG = 13

## OFFSET EEPROM ADDRESS
LINEAR_INTERCEPT_START_ADDRESS = 70
LINEAR_SLOPE_START_ADDRESS = 50
OFFSET_START_ADDRESS = 30
OFFSET_STRETCH_START_ADDRESS = 20

SERIAL_NUMBER_ADDRESS = 100

## EEPROM DATA TYPE INDEX
EEPROM_DATA_TYPE_BYTE = 1
EEPROM_DATA_TYPE_INTEGER = 2
EEPROM_DATA_TYPE_FLOAT = 4

## PROTOCOL MESSAGE
# READY & Response
READY                   = "@1"
OK                      = "OK"
# Set Command
SET_POSITION            = "G0 X{} Y{} Z{} F{}"
SET_POSITION_RELATIVE   = "G2204 X{} Y{} Z{} F{}"
SET_SERVO_ANGLE         = "G2202 N{} V{}"
STOP_MOVING             = "G2203"
SET_PUMP                = "M2231 V{}"
GET_PUMP                = "P2231"
SET_GRIPPER             = "M2232 V{}"
SET_BUZZER              = "M2210 F{} T{}"
SET_POLAR               = "G2201 S{} R{} H{} F{}"
ATTACH_SERVO            = "M2201 N{}"
DETACH_SERVO            = "M2202 N{}"
# Get Command
GET_SIMULATION          = "M2222 X{} Y{} Z{} P0"
GET_FIRMWARE_VERSION    = "P2203"
GET_HARDWARE_VERSION    = "P2202"
GET_COOR                = "P2220"
GET_SERVO_ANGLE         = "P2200"
GET_IS_MOVE             = "M2200"
GET_TIP_SENSOR          = "P2233"
GET_POLAR               = "P2221"
GET_GRIPPER             = "P2232"
GET_EEPROM              = "M2211 N0 A{} T{}"
SET_EEPROM              = "M2212 N0 A{} T{} V{}"
GET_ANALOG              = "P2241 N{}"
GET_DIGITAL             = "P2240 N{}"

# Report Command
SET_REPORT_POSITION     = "M120 V{}"
REPORT_POSITION_PREFIX  = "@3"

# Protocol Select
SET_PROTOCOL = "FOO"
